### PR TITLE
fix: callback for `mobius_matvec` for higher dimensions

### PR DESCRIPTION
**Pull request by GitZH-Chen**
solution for the issue #233:

Bug: Incorrect use of `.prod(..., dtype=torch.bool)` in `_mobius_matvec` causes invalid fallback condition in high dimensions

Description:

In the current implementation of `_mobius_matvec`, the fallback condition is written as:

    cond = (mx == 0).prod(dim=dim, keepdim=True, dtype=torch.bool)

This line can incorrectly return `True` even when all entries of `mx` are non-zero, especially when `mx` has large dimensions (e.g., 128 or more). This leads to incorrect output by triggering the fallback case and selecting `res_0` instead of the correct Möbius result `res_c`.

How to Reproduce:

    import torch

    mx = torch.randn(2665, 128, dtype=torch.float64)
    print((mx == 0).sum())  # Expect 0
    cond = (mx == 0).prod(dim=1, keepdim=True, dtype=torch.bool)
    print(cond.unique())    # Expected: tensor([False]); Got: tensor([True]) ❌

Root Cause:

`.prod(..., dtype=torch.bool)` on boolean tensors is not a safe replacement for logical `.all()`. PyTorch may internally interpret this as a numeric product followed by a type cast, and in high dimensions this leads to incorrect `True` outputs.

Recommended Fix:

Replace:

    cond = (mx == 0).prod(dim=dim, keepdim=True, dtype=torch.bool)

with:

    cond = (mx == 0).all(dim=dim, keepdim=True)

This correctly returns `True` only when all elements along the dimension are exactly zero.

Suggested Fix Implementation:

    @torch.jit.script
    def _mobius_matvec(m: torch.Tensor, x: torch.Tensor, k: torch.Tensor, dim: int = -1):
        if m.dim() > 2 and dim != -1:
            raise RuntimeError("broadcasted Möbius matvec is supported for the last dim only")
        x_norm = x.norm(dim=dim, keepdim=True, p=2).clamp_min(1e-15)
        if dim != -1 or m.dim() == 2:
            mx = torch.tensordot(x, m, ([dim], [1]))
        else:
            mx = torch.matmul(m, x.unsqueeze(-1)).squeeze(-1)
        mx_norm = mx.norm(dim=dim, keepdim=True, p=2).clamp_min(1e-15)
        res_c = tan_k(mx_norm / x_norm * artan_k(x_norm, k), k) * (mx / mx_norm)
        cond = (mx == 0).all(dim=dim, keepdim=True)  # ✅ fixed condition
        res_0 = torch.zeros(1, dtype=res_c.dtype, device=res_c.device)
        res = torch.where(cond, res_0, res_c)
        return res

### CODE REVIEW

**Review by SauravMaheshkar — CHANGES_REQUESTED** [review]
Could you add a test to confirm the intended behavior as well.

**Comment by GitZH-Chen** [comment]
> Could you add a test to confirm the intended behavior as well.

Sure! I’ve added the following test into the end of `test_gyrovector_math.py`:

```python
def test_mobius_matvec_vs_log_exp_equivalence():
    """
    Test: mobius_matvec ≈ expmap0(logmap0 @ W)
    Now works for high dimensions (e.g., 128), previously only stable for small dims (e.g., 20).
    """
    from geoopt.manifolds import Stereographic
    import torch as th

    dtype = th.float64
    manifold = Stereographic(k=2.0).to(dtype)

    batch_size = 100
    dim = 128
    dim_out = 128

    x = manifold.random_normal(batch_size, dim).to(dtype)
    weight = th.randn(dim_out, dim, dtype=dtype)

    res1 = manifold.mobius_matvec(weight, x)
    res2 = manifold.expmap0(th.tensordot(manifold.logmap0(x), weight, dims=([-1], [1])))

    max_diff = (res1 - res2).abs().max()
    print(f"[✓] Max absolute difference: {max_diff:.3e}")

    assert max_diff < 1e-5

File: .DS_Store
Comment (SauravMaheshkar) [inline .DS_Store:1]: We shouldn't add auto generated MacOS files. Maybe you could add `**/.DS_Store` to `.gitignore`

File: geoopt/manifolds/stereographic/math.py [hunk 1]
```diff
@@ -1233,7 +1233,9 @@ def _mobius_matvec(m: torch.Tensor, x: torch.Tensor, k: torch.Tensor, dim: int =
         mx = torch.matmul(m, x.unsqueeze(-1)).squeeze(-1)
     mx_norm = mx.norm(dim=dim, keepdim=True, p=2).clamp_min(1e-15)
     res_c = tan_k(mx_norm / x_norm * artan_k(x_norm, k), k) * (mx / mx_norm)
-    cond = (mx == 0).prod(dim=dim, keepdim=True, dtype=torch.bool)
+    # this could be wrong, which is in geoopt:
+    # cond = (mx == 0).prod(dim=dim, keepdim=True, dtype=torch.bool)
```
Comment (SauravMaheshkar) [inline geoopt/manifolds/stereographic/math.py:1237]: we don't need these comments

File: tests/test_gyrovector_math.py [hunk 2]
```diff
@@ -696,3 +696,29 @@ def test_weighted_midpoint_weighted_zero_sum(_k, lincomb):
...
     mid.sum().backward()
     assert torch.isfinite(a.grad).all()
+
+def test_mobius_matvec_vs_log_exp_equivalence():
+    """
+    Test: mobius_matvec = expmap0(W logmap0(x))
+    Now works for high dimensions (e.g., 128), previously only stable for small dims (e.g., 20).
+    """
+    from geoopt.manifolds import Stereographic
+    import torch as th
```
Comment (SauravMaheshkar) [inline tests/test_gyrovector_math.py:706]: Both of these packages are imported at the top of the file, let's use those import aliases

File: tests/test_gyrovector_math.py [hunk 3]
```diff
@@ -696,3 +696,29 @@ def test_weighted_midpoint_weighted_zero_sum(_k, lincomb):
...
+    dim_out = 128
+
+    x = manifold.random_normal(batch_size, dim).to(dtype)
+    weight = th.randn(dim_out, dim, dtype=dtype)
+
+    res1 = manifold.mobius_matvec(weight, x)
+    res2 = manifold.expmap0(th.tensordot(manifold.logmap0(x), weight, dims=([-1], [1])))
+
+    max_diff = (res1 - res2).abs().max()
+    print(f"[✓] Max absolute difference: {max_diff:.3e}")
```
Comment (SauravMaheshkar) [inline tests/test_gyrovector_math.py:722]: we don't have `print` statements in our tests

File: tests/test_gyrovector_math.py [hunk 4]
```diff
@@ -696,3 +696,29 @@ def test_weighted_midpoint_weighted_zero_sum(_k, lincomb):
...
+
+    dtype = th.float64
+    manifold = Stereographic(k=2.0).to(dtype)
+
+    batch_size = 100
+    dim = 128
+    dim_out = 128
+
+    x = manifold.random_normal(batch_size, dim).to(dtype)
+    weight = th.randn(dim_out, dim, dtype=dtype)
```
Comment (SauravMaheshkar) [inline tests/test_gyrovector_math.py:716]: Let's use these values directly

```suggestion
    x = manifold.random_normal(100, 128).to(dtype)
    weight = th.randn(128, 128, dtype=dtype)
```

**Review by SauravMaheshkar — CHANGES_REQUESTED** [review]

File: tests/test_gyrovector_math.py [hunk 5]
```diff
@@ -696,3 +696,22 @@ def test_weighted_midpoint_weighted_zero_sum(_k, lincomb):
     assert torch.isfinite(mid).all()
     mid.sum().backward()
     assert torch.isfinite(a.grad).all()
+
+def test_mobius_matvec_vs_log_exp_equivalence():
+    """
+    Test: mobius_matvec = expmap0(W logmap0(x))
```
Comment (SauravMaheshkar) [inline tests/test_gyrovector_math.py:702]: We don't have comments in our tests

File: .gitignore [hunk 6]
```diff
@@ -114,3 +117,4 @@ testing-report.html
 test-output.xml
 *.el
 .vscode/
+.DS_Store
```
Comment (SauravMaheshkar) [inline .gitignore:120]: This appears to be an extra line

**Review by SauravMaheshkar — CHANGES_REQUESTED** [review]

**Review by SauravMaheshkar — COMMENTED** [review]
The CI is failing, could you format the files using `black` please
